Toggle comment cursor placement

Aerijo opened the pull request:
### Requirements for Adding, Changing, or Removing a Feature

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must contribute a change that has been endorsed by the maintainer team. See details in the template below.
* The pull request must update the test suite to exercise the updated functionality. For guidance, please see https://flight-manual.atom.io/hacking-atom/sections/writing-specs/.
* After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see https://github.com/atom/atom/tree/master/CONTRIBUTING.md#pull-requests.

### Issue or RFC Endorsed by Atom's Maintainers

<!--

Link to the issue or RFC that your change relates to. This must be one of the following:

* An open issue with the `help-wanted` label
* An open issue with the `triaged` label
* An RFC with "accepted" status

To contribute an enhancement that isn't covered by one of the items above, please follow our guide for suggesting an enhancement: https://github.com/atom/atom/blob/master/CONTRIBUTING.md#suggesting-enhancements

To contribute other changes, you must use a different template. You can see all templates at https://github.com/atom/atom/tree/master/.github/PULL_REQUEST_TEMPLATE.

-->

https://github.com/atom/atom/pull/17519 (made this new one because I'd deleted the old PR branch)

### Description of the Change

<!--

We must be able to understand the design of your change from this description. If we can't get a good idea of what the code will be doing from the description here, the pull request may be closed at the maintainers' discretion. Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code here recently, so please walk us through the concepts.

-->

As in the above PR, this changes selection and cursor behaviour when toggling block comments.

### Alternate Designs

An option to do this at the same place that changes the range in the first place would be better, but I don't know how to go about that.

### Possible Drawbacks

It makes the behaviour more complicated, and could have a performance decrease. The behaviour is toggleable though, at the TextEditor level (I did not pass the options received by the Selection method though).

### Verification Process

Manual testing, with all the scenarios laid out in the linked PR.

Also tested with indentation.

### Release Notes

- More intuitive cursor positioning when toggling block comments

Aerijo committed:
Reimplement toggle comment cursor behaviour (and fix bug)

Aerijo committed:
Merge branch 'master' of https://github.com/atom/atom into comment-cursor-placement

Aerijo committed:
Start to add specs (in wrong file though)

Aerijo committed:
See how this goes

Aerijo committed:
Try again

Aerijo committed:
Fix lint

Aerijo committed:
Try new way

thomasjo commented on src/text-editor.js:
```diff
@@ -4794,6 +4794,24 @@ class TextEditor {
+              let endDelta = oldRange.end.column === endLineLength ? [0, -commentEndString.length - 1] : [0, 0]
+              options.selection.setBufferRange(oldRange.translate(startDelta, endDelta), {autoscroll: false})
+            }
+          }
```
All of the `let` statements above can/should be `const`.

thomasjo commented on src/text-editor.js:
```diff
@@ -4794,6 +4794,24 @@ class TextEditor {
+              if (oldRange.start.column === endLineLength) {
+                let endCol = endLineLength - commentEndString.length - 1
+                options.selection.setBufferRange([[end, endCol], [end, endCol]], {autoscroll: false})
+                return
```
Maybe I'm missing something, but as far as I can tell this `return` statement is unnecessary.

Aerijo commented:
This does change functionality for pressing `tab` when text is selected (the failing test). I'll have to remove the `exclusive` flag, which is a little disappointing. Unless that can be special cased? I still think `exclusive` fits better, but I don't know what else it might break.

Aerijo committed:
Remove exclusive

Aerijo committed:
address feedback

Aerijo commented:
There is one undesired behaviour now (that I know of), for three selections where the first goes right to left on the leftmost character (line 13 in first GIF).

It currently will select the start delim, but only when there are three or more of them. It's weird.

Aerijo committed:
make change

Aerijo commented:
FYI @maxbrunsfeld [This test](https://ci.appveyor.com/project/Atom/atom/builds/20463609/job/riruaptl9gl6taym#L107) seems to have randomly failed.

Aerijo committed:
Merge branch 'master' of https://github.com/atom/atom into comment-cursor-placement

Aerijo committed:
Add more tests

Aerijo commented:
@lee-dohm (I moved to a different PR)

I'm pretty happy with the behaviour now. There are technically still more cases that can be tested, but at the same time I don't think we can exhaustively cover them all. The behaviour, except for as note about line 13, is as in the GIFs above.

Commit:
Merge branch 'master' into comment-cursor-placement

nathansobo commented on spec/text-editor-spec.js:
```diff
@@ -7314,6 +7314,41 @@ describe('TextEditor', () => {
         expect(editor.lineTextForBufferRow(0)).toBe('test')
       })
+
+      it('does not select the new delimiters'), () => {
```
Hi. Unfortunately, the closing parentheses after the description string in this test means that the block following the test isn't running at all. So the test is green only because it isn't really running. I'll try and get the test passing, but I may kick this back to you if it ends up being to much work.

Commit:
Fix syntax error that caused test not to run and some problems with test

nathansobo commented:
Okay, I merged in master and fixed a couple problems with the test once I got it actually running. I'd like to stare at this a bit longer and see if we can simplify it at all. Stay tuned.

Commit:
Clean up assertion style

I prefer to express only one assertion per line rather than &&-ing 
together multiple assertions into a condition. I also prefer to use 
equality assertions so that failure messages include more information 
about the actual and expected values.

I used nested scope blocks so we could re-define the `range` constant in 
a local scope without needing to mutate a variable across unrelated 
tests.

nathansobo commented:
Alright. Your earlier comments about using `exclusive` prompted some investigation. As you also seemed to discover, the `exclusive` option on the selection's marker works great unless the selection is empty. In that case, the cursor is still pushed to the right when commenting at the end of the line. We really need a mechanism like a `bias` that causes the endpoints of a marker to prefer to stay left or move right when there is an insertion at their location. It's something we had in Xray. But for the purposes of fixing this issue, I think your approach is reasonable. Just gonna wait on a green build.

Commit:
Fix lint error

Commit:
Unfocus test 🙈

nathansobo commented:
Oh jeez I left a test focused. Rebuilding 🙄.
